### ISSUE TITLE
Support *sending* keepalive requests

### ISSUE THREAD

**Issue by matthijskooijman** [ronf/asyncssh#190]
Since a while, AsyncSSH supports responding to keepalive requests sent by its peer. However, it would also be useful if it could send these requests on its own. There's two parts to it:
 - Send the requests whenever the connection is idle. This keeps the connection alive, in case there are intermediate NAT routers or other devices that do connection tracking.
 - Terminate the connection when the (keepalive) requests are not being answered. This helps detect when the connection (or the peer) is dead.

This is similar to the TCP keepalive that is already enabled on the socket, except that doing the keepalives inside the SSH connection prevents spoofing them. Also, the default TCP keepalive on Linux only starts sending keepalives after 2 hours of inactivity (which probably works well enough to keep intermediate NAT routers awake, but doesn't really help to detect stale connections). Modifying the TCP keepalive timing seems to use some non-portable (Linux-specific) socket APIs, so I'd rather use SSH keepalives instead.

I suspect as single implementation can be used for both client and server connections (e.g. in the `SSHConnection` class?). Configuring this can probably be done using parameters when creating the connection, or by calling some method later. For the actual parameters and underlying behaviour, we can probably copy what OpenSSH does. E.g. from the client config manpage `ssh_config`:

> ServerAliveCountMax

> Sets the number of server alive messages (see below) which may be sent without ssh(1) receiving any messages back from the server. If this threshold is reached while server alive messages are being sent, ssh will disconnect from the server, terminating the session. It is important to note that the use of server alive messages is very different from TCPKeepAlive (below). The server alive messages are sent through the encrypted channel and therefore will not be spoofable. The TCP keepalive option enabled by TCPKeepAlive is spoofable. The server alive mechanism is valuable when the client or server depend on knowing when a connection has become inactive.

> The default value is 3. If, for example, ServerAliveInterval (see below) is set to 15 and ServerAliveCountMax is left at the default, if the server becomes unresponsive, ssh will disconnect after approximately 45 seconds. This option applies to protocol version 2 only.

> ServerAliveInterval

> Sets a timeout interval in seconds after which if no data has been received from the server, ssh(1) will send a message through the encrypted channel to request a response from the server. The default is 0, indicating that these messages will not be sent to the server. This option applies to protocol version 2 only. 

The configuration directives from `sshd_config` are essentially identical.

**Comment by ronf** [comment]
Yeah - this seems like it could be a useful feature. Do you think it would be sufficient to just implement the global request version of this message and not bother with sending it as a channel message? I haven't looked to see exactly when OpenSSH uses one vs. the other, but I have seen both flavors of the message there.

As you said, it should be possible to have a single implementation for both client & server here.

I'll probably make these extra parameters on the create calls. I'm not sure I see much need to let you change the interval or max count of this keepalive on an existing connection, and supporting that would only complicate the implementation.

**Comment by ronf** [comment]
Quick update - I have this working on both the client and server side, controlled by two new arguments called alive_interval and alive_count_max on the create_connection() and create_server() top-level AsyncSSH functions and the various wrappers around them. The meaning of these arguments is pretty much the same as the OpenSSH interval & count max arguments, and I kept the same defaults (interval as 0 so the feature is disabled by default, and count max set to 3).

As mentioned above, I've decided to only implement the global request version of the keepalive for now for both client and server. If there turns out to be some benefit to sending a channel message instead, though, I can look into that in cases where a channel is open.

I need to add unit tests and do a bit more review, but I'm hoping to get this done by the weekend.

**Comment by matthijskooijman** [comment]
> Quick update - I have this working on both the client and server side

Nice! I'm happy to do some review and testing before you merge (but probably not before next week). Just let me know where to look.

> Do you think it would be sufficient to just implement the global request version of this message and not bother with sending it as a channel message? 

I think so, I suspect that OpenSSH might actually do the same. For all usecases I have, checking the TCP connection is the most important. The only reason I can imagine for do channel checks is when a channel somehow has a different endpoint than the connection, but that should not normally happen (I tried to implement something like this once, but it was messy).

> I'll probably make these extra parameters on the create calls. I'm not sure I see much need to let you change the interval or max count of this keepalive on an existing connection, and supporting that would only complicate the implementation.

In my [code](https://github.com/matthijskooijman/nuttssh/blob/master/nuttssh/server.py), I'm using `asyncssh.create_serve()` with a `server_factory` to handle the connection creation. The way I have things now, my code does not actually create the connection, but get it passed to the `connection_made()` method. It would be easiest if I could just enable keepalive on the existing connection there.

However, reading your last comment better, it seems you've already added these arguments to `create_server()`, presumably to be passed along to all created connections, so that would actually solve my case already.

One case I can still imagine is that you want to change keepalive settings based on who is connecting, what commands are running, etc., which would need a method to change the settings. I can imagine this complicates the code somewhat, to reset counters, timers etc.

**Comment by ronf** [comment]
> Nice! I'm happy to do some review and testing before you merge (but probably not before next week). Just let me know where to look.

This is now in the "develop" branch as commit a4b76cf.

> > Do you think it would be sufficient to just implement the global request version of this message and not bother with sending it as a channel message?
> 
> I think so, I suspect that OpenSSH might actually do the same. For all usecases I have, checking the TCP connection is the most important. The only reason I can imagine for do channel checks is when a channel somehow has a different endpoint than the connection, but that should not normally happen (I tried to implement something like this once, but it was messy).

In the case of OpenSSH, the client side seems to always use global requests, but the server side actually prefers to use channel requests when there's a channel open, only using global requests when there are no channels open. When it does use channels, though, it only sends the keepalive on the first open channel. For now, I've chosen not to mimic this behavior.

> However, reading your last comment better, it seems you've already added these arguments to `create_server()`, presumably to be passed along to all created connections, so that would actually solve my case already.

Yes, that's what I originally had in mind.

> One case I can still imagine is that you want to change keepalive settings based on who is connecting, what commands are running, etc., which would need a method to change the settings. I can imagine this complicates the code somewhat, to reset counters, timers etc.

That's a good point. It turned out this wasn't really much additional work, so I've included this option as well. I changed all wording of "alive" to "keepalive", though, as set_keepalive() reads a bit better than set_alive(). So, you now have the option of using the keepalive_interval and keepalive_count_max arguments on the create call or of calling set_keepalive() with interval and/or count_max arguments on the SSHConnection object you get back if you want to set it later. Both arguments to set_keepalive() have a default that leaves the corresponding value unchanged, allowing you to set just the interval or just the max count using keyword argument syntax.

Thanks again for the suggestion, and let me know what you think. I'll wait a bit before I actually roll this into an official release to give you a chance to take a look.

**Comment by matthijskooijman** [comment]
I just gave this a test drive sending keepalives from my server, seems to work fine :-D

I had a quick glance on the code, which also looks good (lot of boilerplate, the core of the change is nice and small).

One remark: When a client is disconnected, the log shows:

    ERROR:root:Connection error: Disconnect Error: Client not responding

However, this does not seem overly clear to me. First, the "Disconnect Error" suggests to me that there was an error while disconnecting, rather than an error that lead to a disconnect. I suspect this an existing wording, technically not caused by this change. Second, the "Client not responding" does not provide any hint that this is related to keepalive, which could perhaps be made more clear?

**Comment by ronf** [comment]
Regarding the error message, I got the wording from OpenSSH -- that outputs something like:

    Timeout, client not responding from user ronf ::1 port 54888

I'm happy to change the AsyncSSH error to be more specific, though. For instance, I could change it to something like "Client not responding to keepalive" or "Client didn't respond to keepalive".

As for the "Disconnect Error" text, you're right that this is not part of this change, but here's an explanation that might help. Just about everything that triggers an unclean close of a connection calls the connection_lost() callback on the SSHClient or SSHServer object with an exception of type asyncssh.DIsconnectError. This could happen as a result of receiving a MSG_DISCONNECT from the remote system or it could happen due to some local error that triggers AsyncSSH to send a MSG_DISCONNECT. See https://asyncssh.readthedocs.io/en/latest/api.html#disconnect-reasons for the full list of possible disconnect reasons (which are defined in the SSH RFC).

These disconnect errors at the connection level get propagated down to the client or server sessions which are still open on that connection. When using the higher-level API, this will generally cause whatever read, drain, or wait call you might be blocked on to unblock and raise that exception.

The DisconnectError exception that's raised contains a code (taken from the list above) and a reason string. In the case of a keepalive failure, the code is DISC_CONNECTION_LOST and reason is "Client not responding", without the "Disconnect Error: " prefix. However, if you convert the exception to text with str(), you'll get the reason prefixed by "Disconnect Error: " as the value returned to distinguish it from other possible exceptions. If you don't want that, you could have your log call refer to exc.reason instead of using exc itself, as long as you know the exception you are trying to print is an AsyncSSH exception and not something local like OSError or ValueError.

As an aside, you can also programmatically test exc.code if you want to different handling depending on the type of disconnect in some cases without having to parse the reason text. Unfortunately, since the code values are set by the SSH RFC, some things get grouped together under a common code like DISC_CONNECTION_LOST or DISC_BY_APPLICATION. That's why there's no keepalive-specific error code here. In such cases, you may need to parse the reason string, but keep in mind that errors generated by the remote system can put whatever they want in their reason string, and it won't be consistent across implementations.

**Comment by matthijskooijman** [comment]
> I'm happy to change the AsyncSSH error to be more specific, though. For instance, I could change it to something like "Client not responding to keepalive" or "Client didn't respond to keepalive".

Either one sounds fine, with a slight preference for `Client not responding to keepalive` here.

> However, if you convert the exception to text with str(), you'll get the reason prefixed by "Disconnect Error: " as the value returned to distinguish it from other possible exceptions.

I wonder if it would be a easy and/or a good idea to change that prefix? Making it "DisconnectError" could actually make it better, since that makes it somewhat more clear that it refers to an exception name. Replacing it by something like "Connection lost" could be even better in terms of log readability, but that loses the link with the actual exception that is being handled.

One alternative thought could be to change the DisconnectError exception name to something like `ConnectionLostError` or just `ConnectionLost`. I do suppose that that name is no longer really appropriate when the exception is first thrown, since at that point it means "The connection should be disconnected", rather than "The connection has been lost". That could also mean that this needs two separate exceptions: One that is thrown when a disconnection needs to happen, and one that is thrown on reads etc. The first can be transformed into the second programmatically, probably just copying the reason string? Not sure how easy this is, and there might be implications for compatibility...

**Comment by ronf** [comment]
> > I'm happy to change the AsyncSSH error to be more specific, though. For instance, I could change it to something like "Client not responding to keepalive" or "Client didn't respond to keepalive".
> 
> Either one sounds fine, with a slight preference for `Client not responding to keepalive` here.

Ok - this change is in.

> One alternative thought could be to change the DisconnectError exception name to something like `ConnectionLostError` or just `ConnectionLost`. I do suppose that that name is no longer really appropriate when the exception is first thrown, since at that point it means "The connection should be disconnected", rather than "The connection has been lost". That could also mean that this needs two separate exceptions: One that is thrown when a disconnection needs to happen, and one that is thrown on reads etc. The first can be transformed into the second programmatically, probably just copying the reason string? Not sure how easy this is, and there might be implications for compatibility...

This is an interesting idea, and I think I might even be able to do it in a backward-compatible way if I make a new set of exceptions which are children of DisconnectError. It'd probably make sense to make a subclass for each disconnect reason, or at least for the ones that are common like ProtocolError and ConnectionLost. This could actually be used to simplify the code where these are raised, as I wouldn't have to pass in the error code explicitly, and I think it should be possible to name things such that different exception names aren't needed when raising vs. catching.

Some extra work would  be required to translate to the new subclasses when a MSG_DISCONNECT is received on the SSH connection, but that shouldn't be hard to do based on the disconnect "code".

For cases where a generic DisconnectError is delivered, putting 'DisconnectError' without a space in the exception text is probably also a worthwhile change, to make it more obvious that's what it is.

 I'll look into this and see if I can make something work here. Thanks for the suggestion!

**Comment by ronf** [comment]
I spent some time today on this, and commit 37cc0b5 is now available in the "develop" branch. It creates subclasses of DisconnectError for all of the common cases, and makes sure that when a MSG_DISCONNECT is received on the SSH connection, it is turned into the correct subclass before being raised. if one exists. Error codes without a custom subclass will still be raised as the parent class (DisconnectError), and in all cases the "code" values are available just as they were before, along with reason and lang.

Due to the inheritance, application code which is using only the older DisconnectError should continue to work fine for both raising and catching errors. It also works fine if errors are raised using the new names but caught as a DisconnectError (possibly testing the code value in the handler to distinguish them).

There is an edge case where an error raised as a DisconnectError may not be properly caught by other code in the same application which uses the new names, but I'm not sure that's worth worrying about. It isn't an issue when receiving errors from a remote system -- they can always be caught as either a DisconnectError or using the new names. It would only be for errors raised and later caught by the same application. If an application is catching errors using the new names, though, I'd expect it to raise them using those names as well.

As part of these changes, I have completely removed the prefix text (e.g. "Disconnect Error", "SFTP Error", etc. that was previously put into the exception text. This text is now just the reason field, and it's up to the caller logging exception text to add whatever they think is needed (if anything) to make it clear what the exception type was.

To that end, I've adjusted the default logging around connection_lost to distinguish between cases where a connection was actually lost at the network level vs. some kind of timeout (keep alive or login timer) vs. some other error which actually caused the connection to be closed. So, the messages that got logged look like:

    Connection closed
    Connection lost
    Client/Server not responding to keepalive
    Connection failed: <reason for the connection being closed>

If you see any logging that needs tweaking after this change, let me know. Overall, I like how this looks, though. The change touched a lot of files/functions, but only because I decided to actively replace all uses of DisconnectError with this cleaner syntax of the new subclasses, both when raising and when catching errors in unit tests (so they are actually more strict about what they expect).

**Comment by ronf** [comment]
Whoops - found a small issue here with a couple of the subclasses I added later in the development. The previous commit left them out of `__init__.py`. Commit c30b13e fixes this.

**Comment by matthijskooijman** [comment]
Looks good, a lot more readable than the old `DisconnectError` everywhere. As for compatibility, I think this should be fine like this?

I haven't tested yet, but based on reading the code, I have a few remarks:

 - When printing exceptions, I see that in `_cleanup()`, the `ConnectionLost` exception is handled specially, by only printing the reason string and not the exception type. I wonder if that special handling is really needed? Maybe it is an indication that the `ConnectionLost` name is not ideal (see also next point)?
 - When seeing "Connection lost", that suggests to me that the TCP connection was unexpectedly lost. However, the two messages that use `ConnectionLost` are "Login timeout expired" and "Server/Client not responding to keepalive", both of which imply that asyncssh decided to terminate the connection (though in both cases it decided to do so assuming that the connection was no longer working properly, so you could say the connection was already lost perhaps).
 - When seeing `connection failed`, that suggests to me that *making* the connection has failed, so no connection has ever existed. That might just be my interpretation, but perhaps something more descriptive could be put there instead? Perhaps `Connection closed: ` is more appropriate? Though that might not be appropriate for all exceptions?

**Comment by ronf** [comment]
> When printing exceptions, I see that in `_cleanup()`, the `ConnectionLost` exception is handled specially, by only printing the reason string and not the exception type. I wonder if that special handling is really needed? Maybe it is an indication that the `ConnectionLost` name is not ideal (see also next point)?
>
> When seeing "Connection lost", that suggests to me that the TCP connection was unexpectedly lost. However, the two messages that use `ConnectionLost` are "Login timeout expired" and "Server/Client not responding to keepalive", both of which imply that asyncssh decided to terminate the connection (though in both cases it decided to do so assuming that the connection was no longer working properly, so you could say the connection was already lost perhaps).

The most common case for ConnectionLost is actually when SSHConnection's connection_lost() method is called, indicating a TCP level connection close or network error of some kind. If an exception is passed in there (such as ConnectionResetError or BrokenPipeError), the message from that exception will be used. If it was a clean TCP connection close but without a proper SSH disconnect message first, the reason text will be set to simply "Connection lost" with no further information.

The other reason to use ConnectionLost as the name here is that the SSH disconnect reason code is named DISC_CONNECTION_LOST, and for the most part I tried to keep my exception names aligned with the SSH naming.

The reason for using ConnectionLost for the other two cases you mentioned is that it's the only disconnect reason that really fits (see https://asyncssh.readthedocs.io/en/latest/api.html#disconnect-reasons for the list). As you said, the keepalive case in particular arguably means the connection really has already been lost and TCP just hasn't detected it yet. The login timeout is harder to justify in that regard, but it still seemed like the closest fit.

Regarding special-casing ConnectionLost's exception text, I really didn't like that much, but I realized when I first began to include the exception name in the message that it was almost always redundant. Adding another prefix just ended up with a log message like "Connection lost: ConnectionLost: Connection lost", which was clearly not good. Even getting rid of the exception name didn't eliminate the redundancy in the case of a clean TCP close, so I thought I'd experiment with using the reason messages directly in that case, without adding a prefix to it.

> When seeing `connection failed`, that suggests to me that _making_ the connection has failed, so no connection has ever existed. That might just be my interpretation, but perhaps something more descriptive could be put there instead? Perhaps `Connection closed: ` is more appropriate? Though that might not be appropriate for all exceptions?

I didn't want to say "closed" here, because I wanted to keep the previous wording I have of 'Connection closed' for the normal SSH clean close case where something like an SSH disconnect "by application" message is received before the TCP connection is closed, or where the local application has called close(). So, I needed another work that wasn't "closed" or "lost" that indicated some kind of failure that led to either AsyncSSH or the remote SSH peer choosing to actively close the connection. Would "failure" here be better than "failed"?

**Comment by matthijskooijman** [comment]
Just a quick note: I read your comment and see the reasoning, but I haven't found the time and concentration to really understand things and respond properly. I'll try to come back to this soon...

**Comment by ronf** [comment]
This feature is now available in 1.16.0. For now, I've adjusted the wording in the "connection failed" case to say "Connection failure" instead.

**Comment by matthijskooijman** [comment]
Thanks!
